### ISSUE TITLE
Python dev headers missing in venv

### ISSUE THREAD

**Issue by arogozhnikov** [astral-sh/uv#9347]
I think python dev headers are now missing when uv installs venv; repro (start from ubuntu 22.04):

confirmed on uv==0.5.0 and 0.5.4, not tested on other versions
```
export $VIRTUAL_ENV=/tmp/.venv
curl -LsSf https://astral.sh/uv/0.5.0/install.sh | sh
. $HOME/.local/bin/env
uv venv --python 3.10 $VIRTUAL_ENV 
. $VIRTUAL_ENV/bin/activate
uv pip install ihm
```

output:

```
[stderr]
src/ihm_format_wrap_1.7.c:203:11: fatal error: Python.h: No such file or directory
  203 | # include <Python.h>
      |           ^~~~~~~~~~
compilation terminated.
error: command '/usr/bin/x86_64-linux-gnu-gcc' failed with exit code 1
  Caused by: This error likely indicates that you need to install a library that provides "Python.h" for ihm@1.7
```

Annoyingly, **it used to work** - I was able to install this package, but a couple of weeks ago binaries shipped with uv were swapped for some other builds, this affected even previous versions of uv.

**Comment by samypr100** [comment]
Similar to https://github.com/astral-sh/uv/issues/8966#issuecomment-2466513707

TL;DR Try [sysconfigpatcher](https://github.com/bluss/sysconfigpatcher) to adjust the uv python downloads to have the proper pkgconfig and sysconfig metadata for your system.

**Comment by zanieb** [comment]
> but a couple of weeks ago binaries shipped with uv were swapped for some other builds, this affected even previous versions of uv.

This doesn't quite make sense, we didn't change previous versions of uv and the Python distributions we use are hard-coded for each release and cannot change.

**Comment by zanieb** [comment]
If you're using a system interpreter, you should install the headers, e.g., the following works:

```
docker run --platform linux/x86_64 -it ghcr.io/astral-sh/uv:bookworm /bin/bash -c "apt update && apt install python3-dev && uv venv && uv pip install ihm"
```

You can also use the system headers for our managed distributions, which surprises me:

```
docker run --platform linux/x86_64 -it ghcr.io/astral-sh/uv:bookworm /bin/bash -c "apt update && apt install clang && uv venv -p 3.10 && uv pip install 'ihm'"
```

I'm not sure if there are downsides to that.

**Comment by arogozhnikov** [comment]
this works:
```
docker run --platform linux/x86_64 -it --rm ghcr.io/astral-sh/uv:bookworm /bin/bash -c "apt update && apt install clang --yes && uv venv -p 3.10 && uv pip install ihm"
```
this doesn't
```
docker run --platform linux/x86_64 -it --rm ghcr.io/astral-sh/uv:bookworm /bin/bash -c "apt update && apt install clang --yes && uv venv -p 3.11 && uv pip install ihm"
```

**Comment by samypr100** [comment]
> Similar to [#8966 (comment)](https://github.com/astral-sh/uv/issues/8966#issuecomment-2466513707)
> 
> TL;DR Try [sysconfigpatcher](https://github.com/bluss/sysconfigpatcher) to adjust the uv python downloads to have the proper pkgconfig and sysconfig metadata for your system.

@arogozhnikov Please use sysconfigpatcher. Your second example does work when you use it.

```bash
docker run --platform linux/x86_64 -it --rm ghcr.io/astral-sh/uv:bookworm /bin/bash -c "apt update && apt install clang --yes && uv venv -p 3.11.10 && uv tool install git+https://github.com/bluss/sysconfigpatcher && uv tool run sysconfigpatcher /root/.local/share/uv/python/cpython-3.11.10-linux-x86_64-gnu && uv pip install ihm"
```

**Comment by wjakob** [comment]
I am also having this issue, and using Docker is not a feasible workaround. Could UV not be modified so that installs the Python headers when creating a venv?

**Comment by samypr100** [comment]
Hi @wjakob, the docker examples here are to provide an MRE, you can apply the same steps outside of docker.

**Comment by wjakob** [comment]
This is on a machine where I don't have root access, hence installing the headers via `apt` is not an option (also, that would be for the OS Python version, which is often quite outdated). With `uv` not installing them as part of the virtual environment, it is impossible to build C extensions within the environment.

**Comment by samypr100** [comment]
> This is on a machine where I don't have root access, hence installing the headers via `apt` is not an option (also, that would be for the OS Python version, which is often quite outdated). With `uv` not installing them as part of the virtual environment, it is impossible to build C extensions within the environment.

Please see https://github.com/astral-sh/uv/issues/9347#issuecomment-2495214667

**Comment by zanieb** [comment]
@arogozhnikov your failing example uses the system interpreter (3.11 is installed in the image already)

```
root@c0d1cea75e52:/# uv python list
cpython-3.13.0+freethreaded-linux-x86_64-gnu    <download available>
cpython-3.12.7-linux-x86_64-gnu                 <download available>
cpython-3.11.10-linux-x86_64-gnu                <download available>
cpython-3.11.2-linux-x86_64-gnu                 /usr/bin/python3.11
cpython-3.11.2-linux-x86_64-gnu                 /usr/bin/python3 -> python3.11
cpython-3.11.2-linux-x86_64-gnu                 /bin/python3.11
cpython-3.11.2-linux-x86_64-gnu                 /bin/python3 -> python3.11
cpython-3.10.15-linux-x86_64-gnu                <download available>
cpython-3.9.20-linux-x86_64-gnu                 <download available>
cpython-3.8.20-linux-x86_64-gnu                 <download available>
cpython-3.7.9-linux-x86_64-gnu                  <download available>
pypy-3.10.14-linux-x86_64-gnu                   <download available>
pypy-3.9.19-linux-x86_64-gnu                    <download available>
pypy-3.8.16-linux-x86_64-gnu                    <download available>
pypy-3.7.13-linux-x86_64-gnu                    <download available>
```

That's why you need to install development headers via the system package manager.

If you force uv to use a managed interpreter with `--python-preference only-manage`, the example works

```
docker run --platform linux/x86_64 -it --rm ghcr.io/astral-sh/uv:bookworm /bin/bash -c "apt update && apt install clang --yes && uv venv -p 3.11 --python-preference only-managed && uv pip install ihm"
```

**Comment by zanieb** [comment]
@wjakob please share a reproducible example

**Comment by samypr100** [comment]
> ```
> docker run --platform linux/x86_64 -it --rm ghcr.io/astral-sh/uv:bookworm /bin/bash -c "apt update && apt install clang --yes && uv venv -p 3.11 --python-preference only-managed && uv pip install ihm"
> ```

Thanks @zanieb, I've been incorrectly under assumption OP wanted to use `gcc` and not `clang`, hence my suggestions to use [sysconfigpatcher](https://github.com/bluss/sysconfigpatcher).

```
docker run --platform linux/x86_64 -it --rm ghcr.io/astral-sh/uv:bookworm /bin/bash -c "apt update && uv venv -p 3.11.10 && uv tool install git+https://github.com/bluss/sysconfigpatcher && uv tool run sysconfigpatcher /root/.local/share/uv/python/cpython-3.11.10-linux-x86_64-gnu && uv pip install ihm"
```

Do you think it's worth expanding in docs when is it reasonable to attempt to use `sysconfigpatcher` by adding a note to `python-versions.md` or maybe a troubleshooting faq.

**Comment by zanieb** [comment]
> Do you think it's worth expanding in docs when is it reasonable to attempt to use sysconfigpatcher by adding a note to python-versions.md or maybe a troubleshooting faq.

It's sort of a priority for me to understand the problems people are encountering and fix it here rather than recommend `sysconfigpatcher` — it's nice that it exists and I'm sure we'll learn from it but I'd prefer a first-party solution.

**Comment by Tristan971** [comment]
Great to see it’s already being looked at! :-)

**Comment by charliermarsh** [comment]
Gonna review this today to see if it's resolved with latest changes.

**Comment by samypr100** [comment]
If it's any help, this would still fail

```
> docker run --platform linux/x86_64 -it --rm ghcr.io/astral-sh/uv:0.5.9-bookworm /bin/bash -c "uv venv -p 3.11.10 && uv pip install ihm"
Using CPython 3.11.10
Creating virtual environment at: .venv
Resolved 2 packages in 330ms
  x Failed to download and build `ihm==1.8`
  `-> Build backend failed to build wheel through `build_wheel` (exit status: 1)

      [stdout]
      running bdist_wheel
      running build
      running build_py
      creating build/lib.linux-x86_64-cpython-311/ihm
      copying ihm/__init__.py -> build/lib.linux-x86_64-cpython-311/ihm
      copying ihm/dataset.py -> build/lib.linux-x86_64-cpython-311/ihm
      copying ihm/representation.py -> build/lib.linux-x86_64-cpython-311/ihm
      copying ihm/analysis.py -> build/lib.linux-x86_64-cpython-311/ihm
      copying ihm/report.py -> build/lib.linux-x86_64-cpython-311/ihm
      copying ihm/metadata.py -> build/lib.linux-x86_64-cpython-311/ihm
      copying ihm/flr.py -> build/lib.linux-x86_64-cpython-311/ihm
      copying ihm/cross_linkers.py -> build/lib.linux-x86_64-cpython-311/ihm
      copying ihm/reader.py -> build/lib.linux-x86_64-cpython-311/ihm
      copying ihm/restraint.py -> build/lib.linux-x86_64-cpython-311/ihm
      copying ihm/source.py -> build/lib.linux-x86_64-cpython-311/ihm
      copying ihm/startmodel.py -> build/lib.linux-x86_64-cpython-311/ihm
      copying ihm/format.py -> build/lib.linux-x86_64-cpython-311/ihm
      copying ihm/model.py -> build/lib.linux-x86_64-cpython-311/ihm
      copying ihm/test.py -> build/lib.linux-x86_64-cpython-311/ihm
      copying ihm/geometry.py -> build/lib.linux-x86_64-cpython-311/ihm
      copying ihm/citations.py -> build/lib.linux-x86_64-cpython-311/ihm
      copying ihm/format_bcif.py -> build/lib.linux-x86_64-cpython-311/ihm
      copying ihm/protocol.py -> build/lib.linux-x86_64-cpython-311/ihm
      copying ihm/multi_state_scheme.py -> build/lib.linux-x86_64-cpython-311/ihm
      copying ihm/dictionary.py -> build/lib.linux-x86_64-cpython-311/ihm
      copying ihm/location.py -> build/lib.linux-x86_64-cpython-311/ihm
      copying ihm/dumper.py -> build/lib.linux-x86_64-cpython-311/ihm
      copying ihm/reference.py -> build/lib.linux-x86_64-cpython-311/ihm
      creating build/lib.linux-x86_64-cpython-311/ihm/util
      copying ihm/util/__init__.py -> build/lib.linux-x86_64-cpython-311/ihm/util
      copying ihm/util/make_mmcif.py -> build/lib.linux-x86_64-cpython-311/ihm/util
      running build_ext
      building 'ihm._format' extension
      creating build/temp.linux-x86_64-cpython-311/src
      clang -pthread -Wsign-compare -Wunreachable-code -DNDEBUG -g -fwrapv -O3
      -Wall -fPIC -fPIC -Isrc -I/root/.cache/uv/builds-v0/.tmpOq7G05/include
      -I/root/.local/share/uv/python/cpython-3.11.10-linux-x86_64-gnu/include/python3.11 -c src/ihm_format.c -o
      build/temp.linux-x86_64-cpython-311/src/ihm_format.o

      [stderr]
      error: command 'clang' failed: No such file or directory
```

**Comment by wjakob** [comment]
@samypr100 You need to have a sane build setup (C++ dev environment) in your container. This is not related to the main issue, which was about the presence of Python headers.

**Comment by samypr100** [comment]
> [@samypr100](https://github.com/samypr100) You need to have a sane build setup (C++ dev environment) in your container. This is not related to the main issue, which was about the presence of Python headers.

Thanks, note this is a `buildpack-deps` based image so it has a sane build setup.

**Comment by zanieb** [comment]
I presume the point is that we'd need #9916 for that example to work?

**Comment by charliermarsh** [comment]
Yeah for me that's separate from this issue (though a good point from @samypr100).

**Comment by charliermarsh** [comment]
Confirmed that this works on latest (with `CC=gcc`).

**Comment by samypr100** [comment]
My last example now also works as of 0.5.10 release without the need for `CC=gcc` 🥳 

```shell
> docker run --platform linux/x86_64 -it --rm ghcr.io/astral-sh/uv:0.5.10-bookworm /bin/bash -c "uv venv -p 3.11.10 && uv pip install ihm"
Using CPython 3.11.10
Creating virtual environment at: .venv
Resolved 2 packages in 406ms
   Built ihm==1.8
Prepared 2 packages in 2.72s
Installed 2 packages in 2ms
 + ihm==1.8
 + msgpack==1.1.0
````

**Comment by jcelerier** [comment]
Hello,
I'm trying to understand in which way this issue is solved.

```
$ docker run --platform linux/x86_64 -it --rm ghcr.io/astral-sh/uv:bookworm /bin/sh -c "uv venv -p 3.13 && find .venv"
Status: Downloaded newer image for ghcr.io/astral-sh/uv:bookworm
Using CPython 3.13.5
Creating virtual environment at: .venv
.venv
.venv/CACHEDIR.TAG
.venv/.gitignore
.venv/bin
.venv/bin/python
.venv/bin/python3
.venv/bin/python3.13
.venv/bin/activate
.venv/bin/activate.csh
.venv/bin/activate.fish
.venv/bin/activate.nu
.venv/bin/activate.ps1
.venv/bin/activate.bat
.venv/bin/deactivate.bat
.venv/bin/pydoc.bat
.venv/bin/activate_this.py
.venv/pyvenv.cfg
.venv/lib
.venv/lib/python3.13
.venv/lib/python3.13/site-packages
.venv/lib/python3.13/site-packages/_virtualenv.py
.venv/lib/python3.13/site-packages/_virtualenv.pth
.venv/lib64
```

At no point the includes are in the venv, they end up in a $HOME/.local/share folder outside of my control.
What must I do to get the python headers of the  version installed by uv in a path accessible from within the .venv ? (or at least a path I can easily access programmatically without knowledge of the user's OS or $HOME).

The context is a C++ software that lives entirely outside of uv's venv but needs to access the Python headers of a python installed with a specific .venv

**Comment by zanieb** [comment]
I don't think it's a part of the standard for includes to be present in the virtual environment. See https://peps.python.org/pep-0405/

> The site and sysconfig standard-library modules are modified such that the standard library and header files are found relative to sys.base_prefix / sys.base_exec_prefix, while site-package directories (“purelib” and “platlib”, in sysconfig terms) are still found relative to sys.prefix / sys.exec_prefix.

Use `sys.base_prefix`?

**Comment by jcelerier** [comment]
I would like to be able to reach these folders without invoking python as part of my cmake scripts - for instance I could be cross-compiling with the python binary not being in the architecture of my host in the part of my build script which needs to know the location of this include folder

**Comment by samypr100** [comment]
> I would like to be able to reach these folders without invoking python as part of my cmake scripts - for instance I could be cross-compiling with the python binary not being in the architecture of my host in the part of my build script which needs to know the location of this include folder

Sadly includes are not in virtualenvs per spec like Zanie mentioned. You can use CMake Python3 [module](https://cmake.org/cmake/help/latest/module/FindPython3.html) for this as suggested in https://github.com/astral-sh/uv/issues/10267#issuecomment-2569938357 and tell it use the virtual environment.

**Comment by jcelerier** [comment]
Thanks, I did not know about these variables, I'll try with them!
